### PR TITLE
storage/kafka: more efficient fetch_partition_info

### PR DESCRIPTION
**Pull request by teskje**
This PR makes the Kafka source's `fetch_partition_info` method use the `offsets_for_times` API instead of `fetch_watermarks`, for topics with more than one partition. The benefit is that the former can fetch offsets for multiple partitions in a single request, whereas the later issues one request per partition. If there are many partitions, `offsets_for_times` should therefore be much more efficient.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [x] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [x] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] If this PR includes major [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note), I have pinged the relevant PM to schedule a changelog post.

### CODE REVIEW

File: src/storage/src/source/kafka.rs [hunk 1]
```diff
@@ -1549,23 +1549,68 @@ mod tests {
 }
 
 /// Fetches the list of partitions and their corresponding high watermark
-fn fetch_partition_info<C: ClientContext>(
-    client: &Client<C>,
+fn fetch_partition_info<C: ConsumerContext>(
+    consumer: &BaseConsumer<C>,
```
Comment (benesch) [inline src/storage/src/source/kafka.rs:1553]: Ha! Using `offsets_for_time` here is clever. I didn't realize that under the hood fetching the offset for the "time" `Offset::End` is exactly equivalent to fetching the high water mark.

File: src/storage/src/source/kafka.rs [hunk 2]
```diff
@@ -1549,23 +1549,68 @@ mod tests {
...
+        let pid = pids[0];
+        let (low, high) = consumer
+            .client()
+            .fetch_watermarks(topic, pid, fetch_timeout)?;
+        let watermarks = WatermarkOffsets {
+            low: low.try_into().expect("invalid negative offset"),
+            high: high.try_into().expect("invalid negative offset"),
+        };
+        return Ok([(pid, watermarks)].into());
+    }
```
Comment (benesch) [inline src/storage/src/source/kafka.rs:1574]: It's a bit of a shame to have to fetch the low water mark at all. We don't use it for anything important, AFAICT. cc @petrosagg to check me on that. Could we just ... not? It seems like it's only used to initialize the start offset of a partition, and that could just as well be set to 0.

Comment (teskje) [reply src/storage/src/source/kafka.rs:1574]: Ha, I was wondering about this! Would be great if we could remove it. Then we'd need fewer requests, and we could throw out the special case here and just always use `offsets_for_time`.

File: src/storage/src/source/kafka.rs (hunk 1 above)
Comment (teskje) [reply src/storage/src/source/kafka.rs:1553]: I figured that out by implementing my own version of `rd_kafka_query_watermark_offsets` that can take multiple partitions and then realizing that I'm just reimplementing `rd_kafka_offsets_for_time` 😅

File: src/storage/src/source/kafka.rs (hunk 2 above)
Comment (petrosagg) [reply src/storage/src/source/kafka.rs:1574]: Looked at the code and I agree that the low watermark could be removed. Let's try it and see what CI thinks of that

**Comment by teskje** [comment]
As far as I can tell all nightlies failures are unrelated, so I think this is ready to review.

**Review by benesch — COMMENTED** [review]
LGTM but deferring formal approval to @petrosagg.

**Review by petrosagg — APPROVED** [review]
LGTM!

**Comment by teskje** [comment]
TFTRs!

**Pull request merged**